pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -886,6 +888,35 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
+    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+        if (!conf.isPerMessageRedeliverUnacknowledged()
+                || messageIds.size() > conf.getMaxRedeliverUnacknowledgedMessagesBatch()) {
+            redeliverUnacknowledgedMessages();
+            return;
+        }
+        ClientCnx cnx = cnx();
+        if (isConnected() && cnx.getRemoteEndpointProtocolVersion() >= ProtocolVersion.v2.getNumber()) {
+            MessageIdData.Builder builder = MessageIdData.newBuilder();
+            List<MessageIdData> messageIdDatas = messageIds.stream()
+                    .map(messageId -> {
+                        builder.setPartition(messageId.getPartitionIndex());
```

Comment:
Shouldn't this be a new builder instance each time the mapping function is called?

Also, the builder itself should be recycled after `builder.build()` is called